hackernews.el

```diff
@@ -353,19 +437,34 @@ N defaults to 1."
 
 (defun hackernews-browse-url-action (button)
   "Pass URL of BUTTON to `browse-url'."
-  (browse-url (button-get button 'shr-url)))
+  (flet ((bget (prop) (button-get button prop)))
+    (let ((id (bget 'id))
+	  (url (bget 'shr-url))
+	  (button-type (bget 'type)))
+      (button-type-put button-type 'hackernews-visited-ids (cons id (button-type-get button-type 'hackernews-visited-ids)))
+      (if hackernews-show-visited-links
+	  (let ((inhibit-read-only t))
+	    (button-put button 'type 'hackernews-link-visited)))
+      (browse-url url))))
 
 (defun hackernews-button-browse-internal ()
   "Open URL of button under point within Emacs.
 The URL is passed to `hackernews-internal-browser-function',
 which see."
   (interactive)
-  (funcall hackernews-internal-browser-function
-           (button-get (point) 'shr-url)))
-
-(defun hackernews--button-string (type label url)
+  (flet ((bget (prop) (button-get (button-at (point)) prop)))
+    (let ((id (bget 'id))
+	  (url (bget 'shr-url))
+	  (button-type (bget 'type)))
+      (button-type-put button-type 'hackernews-visited-ids (cons id (button-type-get button-type 'hackernews-visited-ids)))
+      (if hackernews-show-visited-links
+	  (let ((inhibit-read-only t))
+	    (button-put (button-at (point)) 'type 'hackernews-comment-count-visited)))
+      (funcall hackernews-internal-browser-function url))))
```

Comment:
1. Better to write `(point)` than `(button-at (point))` because we are dealing with text, not overlay, buttons and `button-at` needlessly creates a new unfreed marker each time it is called. See [`(elisp) Overview of Markers`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Overview-of-Markers.html) for what I mean by "freeing" a marker, i.e. making it point to `nil`.
2. This completely duplicates `hackernews-browse-url-action`, why not generalise the two?

```el
(defun hackernews--visit (button fn)
  "Visit URL of BUTTON by passing to to FN."
  (let* ((type  (button-type button))
         (vtype (button-type-get type 'hackernews-visited-button))
         (inhibit-read-only t))
    (when (and hackernews-show-visited-links
               (not (eq type vtype)))
      (button-put button 'type vtype)
      (puthash (button-get button 'id)
               t                        ; TODO: Replace with metadata
               (cdr (assq type hackernews--visited-ids)))))
  (funcall fn (button-get button 'shr-url)))

(defun hackernews-browse-url-action (button)
  "Pass URL of BUTTON to `browse-url'."
  (hackernews--visit button #'browse-url))

(defun hackernews-button-browse-internal ()
  "Open URL of button under point within Emacs.
The URL is passed to `hackernews-internal-browser-function',
which see."
  (interactive)
  (hackernews--visit (point) hackernews-internal-browser-function))
```